### PR TITLE
[New API] add new api paddle.mode and paddle.Tensor.mode

### PR DESCRIPTION
**Pull request by zoooo0820**
### PR types
New features
### PR changes
OPs

### Describe
add new api paddle.mode and paddle.Tensor.mode

mode-API文档-PaddlePaddle深度学习平台 - 10 136 157 23 

mode-API Document-PaddlePaddle Deep Learning Platform - 10 136 157 23

### CODE REVIEW

File: paddle/fluid/operators/mode_op.h [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+      auto e_input = framework::EigenMatrix<T>::Reshape(*input, input_dim - 1);
+      auto e_indices =
+          framework::EigenMatrix<Type>::Reshape(*indices, input_dim - 1);
+      output_data[i * input_width + e_indices(i, 0)] = e_input(i, 0);
+    }
+  }
+}
+
+template <typename DeviceContext, typename T>
+class ModeCPUKernel : public framework::OpKernel<T> {
```
Comment (jeff41404) [inline paddle/fluid/operators/mode_op.h:92]: shall we define ModeCPUKernel and ModeOpCUDAKernel in one class template?

File: paddle/fluid/operators/mode_op.h [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        output->Resize(tmp_out_dims);
+        indices->Resize(tmp_out_dims);
+      }
+
+      // get the trans input_dims, out_dims
+      framework::DDim trans_dims(in_dims);
+      framework::DDim trans_out_dims(in_dims);
+
+      for (size_t i = 0; i < trans.size(); i++) {
+        trans_dims[i] = in_dims[trans[i]];
```
Comment (jeff41404) [inline paddle/fluid/operators/mode_op.h]: the variable name of trans/trans_dims is ambiguous, trans_axis or trans_shape may be clear ?

Comment (zoooo0820) [reply paddle/fluid/operators/mode_op.h]: done

File: paddle/fluid/operators/mode_op.h (hunk 1 above)
Comment (zoooo0820) [reply paddle/fluid/operators/mode_op.h:92]: This will be fixed later

**Review by jeff41404 — APPROVED** [review]
LGTM

**Review by dingjiaweiww — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LG API

**Pull request merged**